content/blog/fp-js-higher-order-functions.md

```diff
@@ -0,0 +1,577 @@
+
+For this example, let's build a small function that works much like an `ifThen` statement. This function should take a boolean `condition` and a function `ifTrue`, that only runs if the `condition` is true.
+
+```js
+function ifThen(condition, ifTrue) {
+  if (condition) {
+    ifTrue()
+  }
+}
+
+ifThen(true, () => console.log("Was true!")) // outputs 'Was true!'
+ifThen(false, () => console.log("Was true!")) // outputs nothing
+```
+
+If you have programmed in javascript for a while, especially node, this will probably be familiar to you. These functions are called `callbacks` and were (and sometimes still are) used in asynchronous functions. They can, however, be used in synchronous functions aswell! An example of a synchronous higher-order function is `addEventListener`, as it takes the function to execute for the event as the second argument.
+
+Let's take a look at the example and see what's going on. The first part with the `if` statement should be pretty clear, we pass in a `boolean` condition and check if it's true. If that's the case, we call the second parameter, `ifTrue` (remember, it's a function).
+
+Let's add an `orElse` that will run if the `condition` is false.
+
+```js
+function ifThen(condition, ifTrue, orElse) {
+  if (condition) {
+    ifTrue()
+  } else {
+    orElse()
+  }
+}
+
+ifThen(
+  true,
+  () => console.log("Was true"),
+  () => console.log("Was false")
+) // Outputs 'Was true'
+ifThen(
+  false,
+  () => console.log("Was true"),
+  () => console.log("Was false")
+) // Outputs 'Was false'
+```
+
+## Example of built-in higher-order functions
+
+Remember when I said that even if you had not heard about higher-order functions before reading this, you had probably already used them anyway. Well, it turns out that javascript has some built-in higher-order functions that operate on arrays. Let's take a look at 3 of them, and why you should use them instead of regular loops.
+
+### Array.prototype.map
+
+`Array.prototype.map` is a function that takes a callback function and runs it over each element in the array, returning a new array. It's great for transforming the values of an array without mutating them.
+
+Let's say you have a shopping cart with some items. These items have a `quantity` and an `itemPrice`, but for presentational purposes, you want to show the total price. Mapping the `products` array to a new one and adding a `totalPrice` property is a perfect usage of the `map` function.
+
+```js
+const products = [
+  {
+    id: 1,
+    name: "Iphone 7",
+    quantity: 1,
+    itemPrice: 1000,
+  },
+  {
+    id: 2,
+    name: "Air Pods",
+    quantity: 1,
+    itemPrice: 200,
+  },
+  {
+    id: 3,
+    name: "Iphone 7 case",
+    quantity: 3,
+    itemPrice: 25,
+  },
+]
+
+function addTotalPriceToProduct(product) {
+  return {
+    ...product,
+    totalPrice: product.itemPrice * product.quantity,
+  }
+}
+
+const productsWithTotalPrice = products.map(addTotalPriceToProduct)
```

Comment:
I would like to see `products.map(product => addTotalPriceToProduct(product))` and then mentioning the short notation